src/MediatR/Mediator.cs

```diff
@@ -73,20 +88,65 @@ public void Publish(INotification notification)
 
         public Task PublishAsync(IAsyncNotification notification)
         {
-            var notificationHandlers = GetNotificationHandlers(notification)
+            var notificationHandlers = GetNotificationHandlers(notification);
+
+            if (_publishOption == PublishAsyncOptions.Parallel)
+            {
+                return PublishParallelAsync(notification, notificationHandlers);
+            }
+            else
+            {
+                return PublishSequentialAsync(notification, notificationHandlers);
+            }
+        }
+
+        public Task PublishAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken)
+        {
+            var notificationHandlers = GetNotificationHandlers(notification);
+
+            if (_publishOption == PublishAsyncOptions.Parallel)
+            {
+                return PublishParallelAsync(notification, cancellationToken, notificationHandlers);
+            }
+            else
+            {
+                return PublishSequentialAsync(notification, cancellationToken, notificationHandlers);
+            }
+        }
+
+        private async Task PublishSequentialAsync(IAsyncNotification notification, IEnumerable<AsyncNotificationHandlerWrapper> notificationHandlers)
+        {
+            foreach (var handler in notificationHandlers)
+            {
+                await handler.Handle(notification);
+            }
+        }
+
+        private Task PublishParallelAsync(IAsyncNotification notification, IEnumerable<AsyncNotificationHandlerWrapper> notificationHandlers)
+        {
+            var tasks = notificationHandlers
                 .Select(handler => handler.Handle(notification))
                 .ToArray();
 
-            return Task.WhenAll(notificationHandlers);
+            return Task.WhenAll(tasks);
         }
 
-        public Task PublishAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken)
+        private Task PublishParallelAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken,
+            IEnumerable<CancellableAsyncNotificationHandlerWrapper> notificationHandlers)
         {
-            var notificationHandlers = GetNotificationHandlers(notification)
+            var tasks = notificationHandlers
                 .Select(handler => handler.Handle(notification, cancellationToken))
                 .ToArray();

```

Comment:
> This is not needed and actually can lead to a wrong behavior.

This is false. [The source](http://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/Task.cs,69351c6da968e5d1) clearly shows that both `WhenAll(IEnumerable<Task>)` and `WhenAll(Task[])` ends up in the same [`InternalWhenAll`](http://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/Task.cs,2529f60968787540) method after checking for `null` tasks.

> Send the IEnumerable directly to Task.WhenAll() and it will do the needful to run them in parallel properly.

As mentioned above, it'll do the exact same thing as doing `ToArray()` before passing it in. It doesn't matter.